engine/src/cubos/engine/tools/scene_editor/plugin.cpp

```diff
@@ -0,0 +1,247 @@
+{
+    ImGui::PushID(&(*scene));
+
+    /// Root node scene
+    bool nodeOpen = ImGui::TreeNode(&(*scene), "%s", scene->name.c_str());
+    scene->isExpanded = nodeOpen;
+
+    // Remove scene
+    if (hierarchyDepth == 1)
+    {
+        ImGui::SameLine();
+        ImGui::PushID(&scene->shouldBeRemoved);
+        if (ImGui::Button("Remove Scene"))
+        {
+            scene->shouldBeRemoved = true;
+        }
+        ImGui::PopID();
+    }
+
+    if (nodeOpen)
+    {
+        if(hierarchyDepth < 2) {
+            char *cstr = new char[64];
+            strcpy_s(cstr, 64, scene->name.c_str());
+
+            ImGui::InputText("##name_change", cstr, 64, ImGuiInputTextFlags_CallbackCharFilter, EntityNameFilter);
+
+            std::string new_name(cstr);
+            if(scene->name.compare(new_name)) {
+                scene->name = new_name;
+            }
+
+            delete[] cstr;
+        }
+        
+        // Add entity to current scene (needs to be root, not a sub scene)
+        if (hierarchyDepth == 0)
+        {
+            if(ImGui::Button("Add Scene")) {
+                std::string sceneName = scene->name + "_scene_" + std::to_string(scene->scenes.size() + 1);
+                SceneInfo newSubscene;
+                newSubscene.name = sceneName;
+                scene->scenes.emplace_back(sceneName, newSubscene);
+            }
+            ImGui::SameLine();
+        }
+        /// Entities
+        drawSceneEntities(scene->entities, scene, selector, cmds, hierarchyDepth);
+
+        /// Scenes
+        std::vector<std::string> sceneNamesToRemove;
+        for (auto& [name, subScene] : scene->scenes)
+        {
+            if (subScene.shouldBeRemoved)
+            {
+                sceneNamesToRemove.push_back(name);
+            }
+            else
+            {
+                drawSceneHierarchy(subScene, cmds, selector, hierarchyDepth + 1);
+            }
+        }
+        for (const auto& s : sceneNamesToRemove)
+        {
+            //scene->scenes.erase(s);
+            CUBOS_TRACE(s);
+        }
+        ImGui::TreePop();
+    }
+
+    ImGui::PopID();
+}
+
+static void sceneEditorSystem(cubos::core::ecs::Commands cmds, Write<SceneInfo> scene, Write<cubos::engine::tools::EntitySelector> selector)
+{
+    ImGui::Begin("Scene Editor");
+    drawSceneHierarchy(scene, cmds, selector, 0);
+    ImGui::End();
+}
+
+void cubos::engine::tools::sceneEditorPlugin(Cubos& cubos)
+{
+    cubos.addPlugin(cubos::engine::imguiPlugin);
+    cubos.addPlugin(cubos::engine::tools::entitySelectorPlugin);
+    cubos.addPlugin(cubos::engine::tools::assetExplorerPlugin);
+    cubos.addPlugin(cubos::engine::transformPlugin);
+
+    cubos.addPlugin(assetsPlugin);
+    cubos.addPlugin(scenePlugin);
+
+    cubos.addResource<SceneInfo>();
+
+    cubos.startupSystem(setSettings).tagged("cubos.settings");
+    cubos.system(checkAssetEvent).tagged("cubos.assets");
```

Comment:
Add the `System` suffix to these systems as well, since we have a ton of other functions here. Maybe also put their definitions adjacent to each other, right before this plugin function, so that its clear where the systems are defined.